File: Networking/Minimum_Spanning_Tree/kruskal.py
"""
Implementation of kruskal and learning PEP8
"""
import sys

def kruskal(graph, node_num):
    """Kruskal:
        Add arcs in increasing order of weight as long as they do not form a cycle
    """
    graph = sorted(graph)
    parent = [0]*node_num
    cluster = [-1]*node_num
    
    def find_root(node):
        """find root of cluster in O(1)"""
        return cluster[parent[node]]

    total = 0
    disjoint_set = 1
    mst = []

    # initiat all nodes to tree 0
    # indicate it has not been visited
    # where find_root() returns 0
    cluster[0] = 0

    for arc in graph:
        weight, source, dest = arc
        # not 0 and same root
        if find_root(source)*find_root(dest) != 0 and find_root(source) == find_root(dest):
            continue

        if find_root(source)+find_root(dest) == 0:
            # not in any cluster either
            # create a new cluster
            parent[source] = disjoint_set
            parent[dest] = disjoint_set
            if cluster[disjoint_set] == -1:
                # cluster hasn't been used
                cluster[disjoint_set] = disjoint_set
                disjoint_set += 1
        elif find_root(source)*find_root(dest) == 0:
            # one of them is in a cluster
            # set its root to the cluster
            parent[dest] = parent[source] = max(parent[dest], parent[source])
        else:   # different cluster
            min_no = min(find_root(dest), find_root(source))
            cluster[parent[source]] = cluster[parent[dest]] = min_no

        total += weight
        mst.append([source, dest, weight])

    return mst, total

def main():
    """Test for kruskal algo
        Read in a file for graph
        compute its mst with kruskal
        and print out mst with total length
    """
    # Read in the test cases in form:
    #   V E _
    #   arc_s arc_d arc_weight
    #   ...
    #   ...
    file_name = sys.argv[1]

    with open(file_name) as file:
        lines = file.readlines()
        # As we need to sort arcs with its weight
        # The simpliest way is to save graph in the form
        # (arc_w, arc_s, arc_d), and we could sort it natually
        node_num, _, _ = map(int, lines[0].split())
        graph = []

        for line in lines[1:]:
            source, end, weight = map(int, line.split())
            graph.append((weight, source, end))

    mst, total = kruskal(graph, node_num)
    print("MST with total weight {}".format(total))
    print(mst)

if __name__ == "__main__":
    main()

File: Networking/Minimum_Spanning_Tree/kruskal.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>

using namespace std;

struct arc {
    int src;
    int dest;
    int weight;

    arc(int src, int dest, int weight): src(src), dest(dest), weight(weight) {
    }
};

int kruskal(const vector<arc> graph, vector<arc> &mst, int node_num) {
    int total = 0, clust_no = 1;
    int *parent = new int[node_num]{};
    int *cluster = new int[node_num];
    fill_n(cluster, node_num, -1);
    cluster[0] = 0;

    auto find_root = [parent, cluster](int node)->int {
        return cluster[parent[node]];
    };

    for (arc a : graph) {
        if (find_root(a.src)*find_root(a.dest) != 0
            && find_root(a.src) == find_root(a.dest))
            continue;

        if (find_root(a.src)+find_root(a.dest) == 0) {
            parent[a.src] = parent[a.dest] = clust_no;
            cluster[clust_no] = clust_no;
            clust_no++;
        } else if (find_root(a.src)*find_root(a.dest) == 0) {
            parent[a.src] = parent[a.dest] = max(parent[a.src], parent[a.dest]);
        } else {
            int min_no = min(find_root(a.src), find_root(a.dest));
            cluster[parent[a.src]] = cluster[parent[a.dest]] = min_no;
        }

        total += a.weight;
        mst.push_back(a);
    }

    return total;
}

int main(int argc, char** argv) {
    ifstream in(argv[1]);
    vector<arc> graph;
    vector<arc> mst;

    int node_num, arc_num, dummy;
    int src, dest, weight;

    in >> node_num >> arc_num >> dummy;
    while (in >> src >> dest >> weight)
        graph.emplace_back(src, dest, weight);

    sort(graph.begin(), graph.end(),
        [](const arc& a1, const arc& a2)->bool {
            return a1.weight < a2.weight;
        });

    int total = kruskal(graph, mst, node_num);

    cout << "Total weight: " << total << endl;
    for (arc a : mst)
        cout << "{" << a.src << ", " << a.dest << ", " << a.weight << "}";
    cout << endl;

    return(0);
}
